src/main/java/org/folio/dew/batch/bulkedit/jobs/TenantResolver.java

```diff
@@ -0,0 +1,83 @@
+import org.folio.dew.service.BulkEditProcessingErrorsService;
+import org.folio.dew.service.ConsortiaService;
+import org.folio.spring.FolioExecutionContext;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.stereotype.Component;
+
+import java.util.HashSet;
+import java.util.Set;
+
+@Component
+@RequiredArgsConstructor
+public class TenantResolver {
+
+  private final FolioExecutionContext folioExecutionContext;
+  private final ConsortiaService consortiaService;
+  private final PermissionsValidator permissionsValidator;
+  private final BulkEditProcessingErrorsService bulkEditProcessingErrorsService;
+  private final UserClient userClient;
+
+  public Set<String> getAffiliatedPermittedTenantIds(EntityType entityType, JobExecution jobExecution, String identifierType, Set<String> tenantIds, ItemIdentifier itemIdentifier) {
+    var affiliatedTenants = consortiaService.getAffiliatedTenants(folioExecutionContext.getTenantId(), folioExecutionContext.getUserId().toString());
+    var jobId = jobExecution.getJobParameters().getString(JobParameterNames.JOB_ID);
+    var fileName = FilenameUtils.getName(jobExecution.getJobParameters().getString(FILE_NAME));
+    var affiliatedAndPermittedTenants = new HashSet<String>();
+    for (var tenantId : tenantIds) {
+      if (!affiliatedTenants.contains(tenantId)) {
+        var user = userClient.getUserById(folioExecutionContext.getUserId().toString());
+        var errorMessage = format(getAffiliationErrorPlaceholder(entityType), user.getUsername(),
+          resolveIdentifier(identifierType), itemIdentifier.getItemId(), tenantId);
+        bulkEditProcessingErrorsService.saveErrorInCSV(jobId, itemIdentifier.getItemId(), errorMessage, fileName);
+      } else if (!isBulkEditReadPermissionExists(tenantId, entityType)) {
+        var user = userClient.getUserById(folioExecutionContext.getUserId().toString());
+        var errorMessage = format(getViewPermissionErrorPlaceholder(entityType), user.getUsername(),
+          resolveIdentifier(identifierType), itemIdentifier.getItemId(), tenantId);
+        bulkEditProcessingErrorsService.saveErrorInCSV(jobId, itemIdentifier.getItemId(), errorMessage, fileName);
+      } else {
+        affiliatedAndPermittedTenants.add(tenantId);
+      }
+    }
+    return affiliatedAndPermittedTenants;
+  }
+
+  private boolean isBulkEditReadPermissionExists(String tenantId, EntityType entityType) {
+    try {
+      return permissionsValidator.isBulkEditReadPermissionExists(tenantId, entityType);
+    } catch (FeignException e) {
+      throw new BulkEditException(e.getMessage());
+    }
+  }
+
+  protected String getAffiliationErrorPlaceholder(EntityType entityType) {
+    if (entityType == EntityType.ITEM) {
+      return NO_ITEM_AFFILIATION;
+    }
+    return NO_HOLDING_AFFILIATION;
+  }
+
+  protected String getViewPermissionErrorPlaceholder(EntityType entityType) {
+    if (entityType == EntityType.ITEM) {
```

Comment:
To protect, please, throw unsupported exception for different to items and holdings cases.

Reply:
Updated